### PR TITLE
zarik staging

### PR DESCRIPTION
**Pull request by zarik5**
This is the first commit where c++ code is optional for the client, controlled by the feature `use-cpp`. c++ is still required to use foveated encoding, so c++ is enabled by default.

### CODE REVIEW

File: alvr/client_core/src/graphics/mod.rs [hunk 1]
```diff
@@ -239,6 +304,46 @@ impl GraphicsContext {
...
+        &self,
+        buffer: *const c_void,
+        texture: gl::Texture,
+        render_cb: impl FnOnce(),
+    ) {
+        if !buffer.is_null() {
+            let client_buffer = (self.get_native_client_buffer)(buffer);
+            check_error(&self.gl_context, "get_native_client_buffer");
+
+            let image = (self.create_image)(
```
Comment (The-personified-devil) [inline alvr/client_core/src/graphics/mod.rs:327]: I don't like that we're continually creating and destroying images

Comment (zarik5) [reply alvr/client_core/src/graphics/mod.rs:327]: This is what we always did. Apparently this is the pattern used both by other opengl and vulkan examples. Technically under the hood this is not too bad. Images are just a handle and the memory is just reused from the decoder android buffer.

**Comment by zarik5** [comment]
To be clear, the staging pass is needed to cache the decoder image. We need to cache it because if the android buffer is null (no newer frame ready) we still need to print something to screen (the last image) otherwise openxr will not like us stalling the render loop and it will crash. Other solutions also don't work, like copying to-from openxr swapchain targets because we need to return used images to the runtime asap, same for decoder buffers.

File: alvr/client_core/src/graphics/mod.rs (hunk 1 above)
Comment (zarik5) [reply alvr/client_core/src/graphics/mod.rs:327]: @barnabwhy there is no image to clear, the memory keeps being swapped

**Review by The-personified-devil — APPROVED** [review]

File: alvr/client_core/src/graphics/mod.rs (hunk 1 above)
Comment (The-personified-devil) [reply alvr/client_core/src/graphics/mod.rs:327]: It probably doesn't really matter for OpenGL especially, so sure

**Pull request merged**